equinox/nn/_batch_norm.py

```diff
@@ -86,26 +98,37 @@ def __init__(
             64-bit mode.
         """
 
+        if approach is None:
+            warnings.warn('BatchNorm approach is None, defaults to approach="batch"')
+            approach = "batch"
+
+        valid_approaches = {"batch", "ema"}
+        if approach not in valid_approaches:
+            raise ValueError(f"approach must be one of {valid_approaches}")
+        self.approach = approach
+
         if channelwise_affine:
             self.weight = jnp.ones((input_size,))
             self.bias = jnp.zeros((input_size,))
         else:
             self.weight = None
             self.bias = None
-        self.first_time_index = StateIndex(jnp.array(True))
+        self.count_index = StateIndex(jnp.array(0, dtype=jnp.int32))
         if dtype is None:
             dtype = default_floating_dtype()
         init_buffers = (
-            jnp.empty((input_size,), dtype=dtype),
-            jnp.empty((input_size,), dtype=dtype),
+            jnp.zeros((input_size,), dtype=dtype),
+            jnp.zeros((input_size,), dtype=dtype),
         )
         self.state_index = StateIndex(init_buffers)
+        self.zero_frac_index = StateIndex(jnp.array(1.0, dtype=dtype))
         self.inference = inference
         self.axis_name = axis_name
         self.input_size = input_size
         self.eps = eps
         self.channelwise_affine = channelwise_affine
         self.momentum = momentum
+        self.warmup_period = max(1, warmup_period)
 
     @jax.named_scope("eqx.nn.BatchNorm")
     def __call__(
```

Comment:
It's not completely obvious to me that the `ema` implementation, with default arguments, reproduces the previous behaviour. (For example, we have `warmup_period=1000` by default?)

Can you add some comments explaining what each `approach` corresponds to?

Reply:
ema with warmup_period=1 approximately reproduces previous behavior.  As I noted the start is different because of how the running statistics are initially populated.  With warmup_period=1 there's no interpolation between the batch and running stats - the running stats are always used as with the previous behavior.  I can give an exact replication with an extra approach if necessary.  

Added some to the documentation

Reply:
I think an exact replication is probably important for the default behaviour, just because I'd like to be sure that we're bit-for-bit backward compatible.

Reply:
Makes sense, it was different enough that I added it as `"ema_compatibility"`.  I changed the warning to rather strongly recommend against using `"ema_compatibility"`.  I haven't found a use case where I wouldn't expect to see the instability (at least with a larger learning rate) but that could very much be due to a lack of imagination on my part..  That part can definitely change if needed